faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+# Copyright 2018-2021 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Interact with server agent.
+"""
+from contextlib import contextmanager
+import json
+from enum import Enum
+
+from attr import attrs, attrib
+from marshmallow import fields, post_load
+from marshmallow_enum import EnumField
+
+from faculty.clients.base import BaseSchema, BaseClient
+
+SERVER_RESOURCES_EVENT = "@SSE/SERVER_RESOURCES_UPDATED"
+
+class ExecutionStatus(Enum):
+    """The status of an environment execution."""
+    NOT_STARTED = "NOT_STARTED"
+    STARTED = "STARTED"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+class EnvironmentExecutionStepStatus(Enum):
+    """The status of an environment execution step."""
+    QUEUED = "QUEUED"
+    CANCELLED = "CANCELLED"
+    STARTED = "STARTED"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+@attrs
+class Execution(object):
+    """Server environment execution.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the execution.
+    status : enum.Enum
```

Comment:
We can be more specific about the type returned here:

```suggestion
    status : ExecutionStatus
```